libosrmc/osrmc.cc

```diff
@@ -245,14 +256,40 @@ float osrmc_table_response_duration(osrmc_table_response_t response, unsigned lo
 
   auto& durations = response_typed->values["durations"].get<osrm::json::Array>();
   auto& durations_from_to_all = durations.values.at(from).get<osrm::json::Array>();
-  auto duration = durations_from_to_all.values.at(to).get<osrm::json::Number>().value;
+  auto optional = durations_from_to_all.values.at(to);
+
+  if (optional.is<osrm::json::Null>()) {
+    *error = new osrmc_error{"NoRoute: Impossible route between points"};
+    return INFINITY;
+  }
+  auto duration = optional.get<osrm::json::Number>().value;
 
   return duration;
 } catch (const std::exception& e) {
   *error = new osrmc_error{e.what()};
   return INFINITY;
 }
 
+float osrmc_table_response_distance(osrmc_table_response_t response, unsigned long from, unsigned long to,
+                                    osrmc_error_t* error) try {
+  auto* response_typed = reinterpret_cast<osrm::json::Object*>(response);
+
+  auto& distances = response_typed->values["distances"].get<osrm::json::Array>();
+  auto& distances_from_to_all = distances.values.at(from).get<osrm::json::Array>();
+  auto optional = distances_from_to_all.values.at(to);
+
+  if (optional.is<osrm::json::Null>()) {
+    *error = new osrmc_error{"NoRoute: Impossible route between points"};
+    return INFINITY;
+  }
+  auto distance = optional.get<osrm::json::Number>().value;
+
+  return distance;
+} catch (const std::exception& e) {
+  *error = new osrmc_error{e.what()};
+  return INFINITY;
+}
+
```

Comment:
What happens if the user calls this function when the distance annotations were not set? Then the JSON "distances" lookup will probably fail throwing an exception? I think we should instead assert this, since it's a user code error calling the distance function without setting distance annotations.

Reply:
I've updated it to assert, but it's a bit ugly - we need to keep a copy of the annotations around in order to assert. I'd love for `libosrm` to return something more structured than JSON, and to push this error handling in there.